File: archive/p/php/fraction-math.php
<?php
function usage()
{
    exit("Usage: ./fraction-math operand1 operator operand2");
}

function parse_int($str_value)
{
    // Remove leading and trailing spaces
    $str_value = trim($str_value);

    // Make sure all digits
    if (preg_match("/^[+-]?\d+$/", $str_value) === FALSE)
    {
        return FALSE;
    }

    // Make sure valid integer
    if (
        filter_var(
            $str_value,
            FILTER_VALIDATE_INT,
            array(
                'options' => array(
                    'decimal' => TRUE,
                    'min_range' => PHP_INT_MIN,
                    'max_range' => PHP_INT_MAX
                )
            )
        ) === FALSE
    )
    {
        return FALSE;
    }

    return intval($str_value);
}

function parse_fraction($str_value)
{
    // Split numerator and denonimator
    $str_parts = explode("/", $str_value, 2);

    // Parse numerator. Exit if numerator invalid
    $num = parse_int($str_parts[0]);
    if ($num === FALSE)
    {
        return FALSE;
    }

    // Assume denominator is 1
    $den = 1;

    // If there is denominator,
    if (count($str_parts) > 1)
    {
        // Parse denominator. Exit if denominator invalid
        $den = parse_int($str_parts[1]);
        if ($den === FALSE)
        {
            return FALSE;
        }
    }

    return new Fraction($num, $den);
}

class Fraction
{
    public function __construct(int $num, int $den)
    {
        $this->num = $num;
        $this->den = $den;
    }

    // Fraction addition
    // n1/d1 + n2/d2 = (n1*d2 + n2*d1) / (d1*d2)
    public function add(Fraction $other): Fraction
    {
        return reduce(
            $this->num * $other->den + $other->num * $this->den,
            $this->den * $other->den
        );
    }

    // Fraction subtraction
    // n1/d1 - n2/d2 = (n1*d2 - n2*d1) / (d1*d2)
    public function sub(Fraction $other): Fraction
    {
        return reduce(
            $this->num * $other->den - $other->num * $this->den,
            $this->den * $other->den
        );
    }

    // Fraction multiplication
    // n1/d1 * n2/d2 = (n1*n2) / (d1*d2)
    public function mult(Fraction $other): Fraction
    {
        return reduce($this->num * $other->num, $this->den * $other->den);
    }

    // Fraction division
    // (n1/d1) / (n2/d2) = (n1*d2) / (n2*d1)
    public function div(Fraction $other): Fraction
    {
        return reduce($this->num * $other->den, $other->num * $this->den);
    }

    // Fraction comparison
    // n1/d1 OP n2/d2 = n1*d2 OP n2*d1
    public function cmp(Fraction $other): int
    {
        return $this->num * $other->den - $other->num * $this->den;
    }

    // Show fraction
    public function show()
    {
        echo "{$this->num}/{$this->den}\n";
    }
}

// Fraction reduction
function reduce($num, $den)
{
    if ($den == 0)
    {
        die("Division by 0");
    }

    // Reduce by dividing numerator and denominator by greatest common denominator,
    // and adjust sign of numerator and denominator as follows:
    //
    // n  d  sign n  sign d
    // +  +     +       +
    // +  -     -       +
    // -  +     -       +
    // -  -     +       +
    $den_sign = ($den > 0) ? 1 : -1;
    $g = gcd($num, $den);
    return new Fraction(intdiv($den_sign * $num, $g), intdiv(abs($den), $g));
}

// Greatest common denominator
function gcd($a, $b)
{
    $a = abs($a);
    $b = abs($b);
    while ($b != 0)
    {
        $t = $b;
        $b = $a % $b;
        $a = $t;
    }

    return $a;
}

// Do fraction math
function fraction_math(Fraction $f1, Fraction $f2, string $op): Fraction|bool
{
    switch ($op)
    {
        case "+":
            return $f1->add($f2);
        case "-":
            return $f1->sub($f2);
        case "*":
            return $f1->mult($f2);
        case "/":
            return $f1->div($f2);
        case ">":
            return $f1->cmp($f2) > 0;
        case ">=":
            return $f1->cmp($f2) >= 0;
        case "<":
            return $f1->cmp($f2) < 0;
        case "<=":
            return $f1->cmp($f2) <= 0;
        case "==":
            return $f1->cmp($f2) == 0;
        case "!=":
            return $f1->cmp($f2) != 0;
        default:
            die("Invalid operator ${op}");
    }
}

// Show fraction result
function show_fraction_result(Fraction|int $value)
{
    if (gettype($value) == "object")
    {
        $value->show();
    }
    else
    {
        echo "${value}\n";
    }
}

// Exit if too few arguments or 2nd argument is empty
if (count($argv) < 4 || empty($argv[2]))
{
    usage();
}

// Parse 1st and 3rd arguments. Exit if invalid
$f1 = parse_fraction($argv[1]);
$f2 = parse_fraction($argv[3]);
if ($f1 === FALSE || $f2 == FALSE)
{
    usage();
}

// Do fraction math and show result
$op = $argv[2];
$result = fraction_math($f1, $f2, $op);
show_fraction_result($result);

File: archive/p/pascal/fraction_math.pas
program FractionMath;

{$mode objfpc}{$H+}{$modeswitch advancedrecords}

uses
   Math,
   SysUtils;

type

   { TFraction }

   TFraction = record
      Numerator, Denominator: int64;

      procedure Reduce;
      function ToString: string;
   class operator +(const A, B: TFraction): TFraction;
   class operator -(const A, B: TFraction): TFraction;
   class operator *(const A, B: TFraction): TFraction;
   class operator /(const A, B: TFraction): TFraction;
   class operator =(const A, B: TFraction): boolean;
   class operator <>(const A, B: TFraction): boolean;
   class operator <(const A, B: TFraction): boolean;
   class operator <=(const A, B: TFraction): boolean;
   class operator >(const A, B: TFraction): boolean;
   class operator >=(const A, B: TFraction): boolean;
   end;

procedure TFraction.Reduce; inline;
var
   Divisor, GcdA, GcdB, Temp: int64;
begin
   if Denominator = 0 then
      raise Exception.Create('Denominator cannot be zero');

   if Numerator = 0 then
   begin
      Denominator := 1;
      Exit;
   end;

   if Denominator < 0 then
   begin
      Denominator := -Denominator;
      Numerator := -Numerator;
   end;

   GcdA := Abs(Numerator);
   GcdB := Denominator;

   while GcdB <> 0 do
   begin
      Temp := GcdB;
      GcdB := GcdA mod GcdB;
      GcdA := Temp;
   end;

   Divisor := GcdA;

   // If the GCD is not 1, then simplify the top and the bottom.
   if Divisor > 1 then
   begin
      Numerator := Numerator div Divisor;
      Denominator := Denominator div Divisor;
   end;
end;

function TFraction.ToString: string;
begin
   Result := Format('%d/%d', [Numerator, Denominator]);
end;

class operator TFraction.+(const A, B: TFraction): TFraction;
begin
   Result.Numerator := A.Numerator * B.Denominator + B.Numerator * A.Denominator;
   Result.Denominator := A.Denominator * B.Denominator;
   Result.Reduce;
end;

class operator TFraction.-(const A, B: TFraction): TFraction;
begin
   Result.Numerator := A.Numerator * B.Denominator - B.Numerator * A.Denominator;
   Result.Denominator := A.Denominator * B.Denominator;
   Result.Reduce;
end;

class operator TFraction.*(const A, B: TFraction): TFraction;
begin
   Result.Numerator := A.Numerator * B.Numerator;
   Result.Denominator := A.Denominator * B.Denominator;
   Result.Reduce;
end;

class operator TFraction./(const A, B: TFraction): TFraction;
begin
   if B.Numerator = 0 then
      raise Exception.Create('Division by zero fraction');

   Result.Numerator := A.Numerator * B.Denominator;
   Result.Denominator := A.Denominator * B.Numerator;
   Result.Reduce;
end;

class operator TFraction.=(const A, B: TFraction): boolean;
begin
   Result := A.Numerator * B.Denominator = B.Numerator * A.Denominator;
end;

class operator TFraction.<>(const A, B: TFraction): boolean;
begin
   Result := not (A = B);
end;

class operator TFraction.<(const A, B: TFraction): boolean;
begin
   Result := A.Numerator * B.Denominator < B.Numerator * A.Denominator;
end;

class operator TFraction.>=(const A, B: TFraction): boolean;
begin
   Result := not (A < B);
end;

class operator TFraction.>(const A, B: TFraction): boolean;
begin
   Result := B < A;
end;

class operator TFraction.<=(const A, B: TFraction): boolean;
begin
   Result := not (B < A);
end;

function ParseFraction(const S: string): TFraction;
var
   Parts: TStringArray;
begin
   Parts := S.Trim.Split(['/']);
   if Length(Parts) <> 2 then
      raise Exception.CreateFmt('Invalid fraction format "%s". Expected "num/den"', [S]);

   if not TryStrToInt64(Parts[0].Trim, Result.Numerator) then
      raise Exception.Create('Invalid numerator');

   if not TryStrToInt64(Parts[1].Trim, Result.Denominator) then
      raise Exception.Create('Invalid denominator');

   if Result.Denominator = 0 then
      raise Exception.Create('Denominator cannot be zero');

   Result.Reduce;
end;

procedure ShowUsage;
begin
   Writeln('Usage: ./fraction-math operand1 operator operand2');
   Halt(1);
end;

function PerformOperation(const Left: TFraction; const Op: string;
const Right: TFraction): string;
begin
   case Op of
      '+': Result := (Left + Right).ToString;
      '-': Result := (Left - Right).ToString;
      '*': Result := (Left * Right).ToString;
      '/': Result := (Left / Right).ToString;
      '==': Result := BoolToStr(Left = Right, '1', '0');
      '!=': Result := BoolToStr(Left <> Right, '1', '0');
      '<': Result := BoolToStr(Left < Right, '1', '0');
      '<=': Result := BoolToStr(Left <= Right, '1', '0');
      '>': Result := BoolToStr(Left > Right, '1', '0');
      '>=': Result := BoolToStr(Left >= Right, '1', '0');
      else
         raise Exception.CreateFmt('Unknown operator: "%s"', [Op]);
   end;
end;


var
   Op: string;
   LeftOperand, RightOperand: TFraction;
begin
   if ParamCount <> 3 then
      ShowUsage;

   try
      LeftOperand := ParseFraction(ParamStr(1));
      Op := ParamStr(2).Trim;
      RightOperand := ParseFraction(ParamStr(3));

      Writeln(PerformOperation(LeftOperand, &Op, RightOperand));
   except
      on E: Exception do
         ShowUsage;
   end;

end.
